notebooks/work-in-progress/ferc1-transform-refactor/ferc1_transform_refactor.py

```diff
@@ -0,0 +1,429 @@
+class TransformParams(BaseModel):
+    """An immutable base model for transformation parameters."""
+
+    class Config:
+        """Prevent parameters from changing part way through."""
+
+        allow_mutation = False
+
+
+class MultiColumnTransformParams(TransformParams):
+    """Transform params that apply to several columns in a table.
+
+    The keys are column names, and the values must all be the same type of
+    :class:`TransformParams` object, since MultiColumnTransformParams are used by
+    :class:`MultiColumnTransformFn` callables.
+
+    Individual subclasses are dynamically generated for each multi-column transformation
+    specified within a :class:`TableTransformParams` object.
+
+    """
+
+    @root_validator
+    def single_param_type(cls, params):  # noqa: N805
+        """Check that all TransformParams in the dictionary are of the same type."""
+        param_types = {type(params[col]) for col in params}
+        if len(param_types) > 1:
+            raise ValueError(
+                "Found multiple parameter types in multi-column transform params: "
+                f"{param_types}"
+            )
+        return params
+
+
+class RenameColumns(TransformParams):
+    """A dictionary for mapping old column names to new column names in a dataframe."""
+
+    columns: dict[str, str] = {}
+
+
+class StringCategories(TransformParams):
+    """Defines mappings to clean up manually categorized freeform strings.
+
+    Each key in a stringmap is a cleaned output category, and each value is the set of
+    all strings which should be replaced with associated clean output category.
+
+    """
+
+    categories: dict[str, set[str]]
+
+    @property
+    def mapping(self) -> dict[str, str]:
+        """A 1-to-1 mapping appropriate for use with :meth:`pd.Series.map`."""
+        return {
+            string: cat for cat in self.categories for string in self.categories[cat]
+        }
+
+
+class UnitConversion(TransformParams):
+    """A column-wise unit conversion."""
+
+    multiplier: float = 1.0
+    adder: float = 0.0
+    pattern: typing.Pattern
+    repl: str
+
+
+class TableTransformParams(TransformParams):
+    """All defined transformation parameters for a table."""
+
+    class Config:
+        """Only allow the known table transform params."""
+
+        extra = "forbid"
+
+    rename_columns: RenameColumns = {}
+    convert_units: dict[str, UnitConversion] = {}
+    categorize_strings: dict[str, StringCategories] = {}
+
+    @classmethod
+    def from_id(cls, table_id: str) -> "TableTransformParams":
+        """A factory method that looks up transform parameters based on table_id."""
+        return cls(**TRANSFORM_PARAMS[table_id])
+
+
+################################################################################
+# Column, MultiColumn, and Table Transform Functions
+################################################################################
+class ColumnTransformFn(Protocol):
```

Comment:
Just curious, why did you decide against having one class for each transform that stores the params and transform logic?

Reply:
I'm not sure exactly what you're imagining, but I wanted to keep the parameters separate from the procedure because we can use exactly the same procedure over and over again, but with different data inputs and different parameters. E.g. the mechanics for categorizing strings or converting units is the same for MWh to KWh as for BTU to mmBTU, but with different parameters controlling the transformation.

I started out defining one `Protocol` for the a mutli-column transform, and another for a column transform, and had classes that implemented both of the Protocols which could transform either a column or all the relevant columns in a table depending on which method & type of parameters it was given. But I wanted to be able to automatically generate the multi-column function from the single-column function using a factory. But with the two separate `Protocols` a class that could do both kinds of transformations would have needed to be able to rely on there being a single column transformation available all the time... which isn't necessarily the case, as there are some table-only transformations.

I guess working with 3 `Protocol`s (column, mutli-column, and table) could fix this.  Every column transform can generate a multi-column transform, and that multi-column transform would implement both the column and table transform protocols.  But there would also be a table transform that is separate and only implements the table transform protocol.

Is it easy to create automatically generate one method in a class from another, factory-style?